metaflow/plugins/kfp/kfp.py

```diff
@@ -28,35 +29,166 @@ def get_ordered_steps(graph):
+        print("_______________ FLOW RUN COMPLETE ________________")
+        return step_output('None', 'None', 'None', 'None', 'None', 'None')
+
+    # TODO: Check why MF echo statements are getting redirected to stderr
+    if len(proc_error) > 1:
+        print("_______________STDERR:_____________________________")
+        print(proc_error)
+
+    if len(proc_output) > 1:
+        print("_______________STDOUT:_____________________________")
+        print(proc_output)
+        # Read in the outputs (in the penultimate line) containing args needed for the next step.
+        # Note: Output format is: ['ds_root', 'run_id', 'next_step', 'next_task_id', 'current_step', 'current_task_id']
+        outputs = (proc_output.split("\n")[-2]).split()
+        print(step_output(outputs[0], outputs[1], outputs[2], outputs[3], outputs[4], outputs[5]))
+    else:
+        raise RuntimeWarning("This step did not generate the correct args for next step to run. This might disrupt "
+                             "the workflow")
+
+    # TODO: Metadata needed for client API to run needs to be persisted outside before this
+    print("--------------- RUNNING: CLEANUP OF TEMP DIR----------")
+    subprocess.call(["rm -r /opt/zillow/.metaflow"], shell=True)
+
+    print("_______________ Done _________________________________")
+    return step_output(outputs[0], outputs[1], outputs[2], outputs[3], outputs[4], outputs[5])
+
+def pre_start_op_func(code_url)  -> NamedTuple('StepOutput', [('ds_root', str), ('run_id', str), ('next_step', str), ('next_task_id', str), ('current_step', str), ('current_task_id', str)]):
+    """
+    Function used to create a KFP container op (see `pre_start_container_op`)that corresponds to the `pre-start` step of metaflow
+
+    """
+
+    import subprocess
+    from collections import namedtuple
+
+    print("\n----------RUNNING: CODE DOWNLOAD from URL---------")
+    subprocess.call(["curl -o helloworld.py {}".format(code_url)], shell=True)
+
+    print("\n----------RUNNING: KFP Installation---------------")
+    subprocess.call(["pip3 install kfp"], shell=True) # TODO: Remove this once KFP is added to dependencies
+
+    print("\n----------RUNNING: METAFLOW INSTALLATION----------")
+    subprocess.call(["pip3 install --user --upgrade git+https://github.com/zillow/metaflow.git@state-integ-s3"],
+                    shell=True)
+
+    print("\n----------RUNNING: MAIN STEP COMMAND--------------")
+    S3_BUCKET = "s3://workspace-zillow-analytics-stage/aip/metaflow"
+    S3_AWS_ARN = "arn:aws:iam::170606514770:role/dev-zestimate-role"
+
+    define_s3_env_vars = 'export METAFLOW_DATASTORE_SYSROOT_S3="{}" && export METAFLOW_AWS_ARN="{}" '.format(S3_BUCKET,
+                                                                                                          S3_AWS_ARN)
+    define_username = 'export USERNAME="kfp-user"'
+    python_cmd = 'python helloworld.py --datastore="s3" --datastore-root="{}" pre-start'.format(S3_BUCKET)
```

Comment:
Can you remind me of what the `pre-start` task actually does for us? Does it create metadata for this specific run?

Reply:
Yes, this is where we get the run_id from which is then used in the other steps.